src/v/io/scheduler.cc

```diff
@@ -0,0 +1,162 @@
+    queue->stop_ = true;
+    queue->sem_.signal();
+    auto monitor = std::exchange(
+      queue->monitor_, seastar::make_ready_future<>());
+    co_await std::move(monitor);
+    co_await queue->io_queue_.stop();
+}
+
+void scheduler::submit_read(queue* queue, page* page) noexcept {
+    queue->io_queue_.submit_read(*page);
+    if (!queue->io_queue_.opened()) {
+        queue->sem_.signal();
+    } else if (queue->cache_hook_.is_linked()) {
+        queue->cache_hook_.unlink();
+        lru_.push_back(*queue);
+    }
+}
+
+void scheduler::submit_write(queue* queue, page* page) noexcept {
+    queue->io_queue_.submit_write(*page);
+    if (!queue->io_queue_.opened()) {
+        queue->sem_.signal();
+    } else if (queue->cache_hook_.is_linked()) {
+        queue->cache_hook_.unlink();
+        lru_.push_back(*queue);
+    }
+}
+
+scheduler::queue::queue(
+  persistence* storage,
+  std::filesystem::path path,
+  io_queue::completion_callback_type complete)
+  : io_queue_(storage, std::move(path), std::move(complete)) {}
+
+const std::filesystem::path& scheduler::queue::path() const noexcept {
+    return io_queue_.path();
+}
+
+// NOLINTNEXTLINE(readability-function-cognitive-complexity)
+seastar::future<> scheduler::monitor(queue* queue) noexcept {
+    while (true) {
+        /*
+         * cooperative scheduling for cache eviction. consider the case when
+         * nofiles_ semaphore is initialized with 0 units (weird, but ok, maybe
+         * we want to be able to pause all I/O). two queues block on nofiles_
+         * semaphore. later a unit is deposited into the semaphore and the first
+         * wakes up and opens its underlying io_queue. at this point the running
+         * queue monitor could be suspended, but we'd like to be able for all
+         * queues to make progress. to handle this the active queue monitor will
+         * evict one or more other queues before suspending itself. however,
+         * because an io queue needs to wake up to be closed there is a delay
+         * before the eviction is visible. this delay is why we track waiters_
+         * separately rather than using the nofiles_ semaphore state.
+         */
+        while (waiters_ > 0 && !lru_.empty()) {
+            waiters_--;
+            auto& queue = lru_.front();
+            lru_.pop_front();
+            queue.sem_.signal();
+        }
+
+        /*
+         * when stopping the queue clean it up fully including removing it from
+         * the cache and closing the underlying io-queue.
+         */
+        if (queue->stop_) {
+            if (queue->io_queue_.opened()) {
+                if (queue->cache_hook_.is_linked()) {
+                    queue->cache_hook_.unlink();
+                }
+                co_await queue->io_queue_.close();
+                queue->units_.return_all();
+            }
+            co_return;
+        }
+
+        co_await queue->sem_.wait(std::max<size_t>(queue->sem_.current(), 1));
```

Comment:
added a comment into the code to explain. basically a condition variable will drop wake-ups if there are no waiters. fix that with a predicate, but in this case the predicate isn't a simple flag or two to check. so we can use a semaphore to ensure that the loop will run again if its signaled while the worker is busy doing other things.

Reply:
Ah got it, thanks for clarifying. A while ago we found that a similar (maybe the same) concern had addressed in seastar recently.

https://github.com/scylladb/seastar/blob/f0b2971a24ab7716933a720572207fa9a1df38bd/tests/unit/condition_variable_test.cc#L44-L49

Is this the scenario you're thinking about?

Reply:
Oh, that's interesting. Seems they addressed that 2 years ago and it hasn't yet solidified in my brain. Part of that is because I don't think pthread condition variable has this property. Anyway, that's good to know. I'll likely switch this back to condition variable then in a later PR.